dev/README_AIRFLOW3_DEV.md

```diff
@@ -0,0 +1,76 @@
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing,
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+-->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of contents**
+
+- [Main branch is Airflow 3](#main-branch-is-airflow-3)
+- [Developing for providers and Helm chart](#developing-for-providers-and-helm-chart)
+- [Developing for Airflow 3](#developing-for-airflow-3)
+- [Developing for Airflow 2.10.x](#developing-for-airflow-210x)
+- [Developing for Airflow 3 and 2.10.x / 2.11.x](#developing-for-airflow-3-and-210x--211x)
+- [Merging PRs 2.10.x](#merging-prs-210x)
+- [Merging PRs for Airflow 3](#merging-prs-for-airflow-3)
+  - [PRs that involve breaking changes](#prs-that-involve-breaking-changes)
+- [Merging PR for Airflow 3 and 2.10.x / 2.11.x](#merging-pr-for-airflow-3-and-210x--211x)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+# Main branch is Airflow 3
+
+The main branch is for development of Airflow 3.
+Airflow 2.10.x releases will be cut from `v-2-10-stable` branch.
+
+# Developing for providers and Helm chart
+
+PRs should target `main` branch.
+
+# Developing for Airflow 3
+
+PRs should target `main` branch.
+
+# Developing for Airflow 2.10.x
+
+PR should target `v-2-10-test` branch. When cutting a new release for 2.10 release manager
+will sync `v-2-10-test`  branch to `v-2-10-stable` and cut the release from the stable branch.
+PR should never target `v-2-10-stable` unless specificly instructed by release manager.
+
+# Developing for Airflow 3 and 2.10.x / 2.11.x
+
+If the PR is relevant for both Airflow 3 and 2 it should target `main` branch.
+
+# Merging PRs 2.10.x
+
+Make sure PR target `v-2-10-test` branch and merge it when ready.
+All regular protocols of merging considerations are applied.
+
+# Merging PRs for Airflow 3
+
+Make sure PR target `main` branch.
+
+## PRs that involve breaking changes
+
+Make sure it has newsfragment, please allow time for community members to review.
+Our goal is to avoid breaking changes if we can so we should allow time to review, don't rush to merge.
+
+# Merging PR for Airflow 3 and 2.10.x / 2.11.x
+
+The committer who merge the PR is responsible for cherry pick the PR to `v-2-10-test`
+If cherry pick is too complex then ask the PR author / start your own PR against `v-2-10-test` directly with the change.
+Note: tracking that the PRs merged as expected is the responsibility of committer who merged the PR.
```

Comment:
```suggestion
The committer who merged the PR is responsible for cherry picking the PR to `v-2-10-test`.
If cherry picking is too complex, ask the PR author / start your own PR against `v-2-10-test` directly with the change.
Note: Tracking that the PRs merged as expected is the responsibility of committer who merged the PR.
```

Nit on syntax.

Does _cherry picking_ in the first line means the committer should just do something like `git cherry-pick <pr-commit> v-2-10-test` and push, or should the cherry-picking change be sent as a PR?

Reply:
I did cherry-picking originall  following the original description but as explained by @ephraimbuddy in  https://apache-airflow.slack.com/archives/C03G9H97MM2/p1723409003786549  - we have "branch-protection" enabled for v2-10-test, so the only way is to make PR against v2-10-test.

@ephraimbuddy @utkarsharma2 -> do I understand correctly? Was that the intention here?

It is quite a burden on the committer merging change (however it makes release manager's job much easier) - and there are pros and cons. 

For one they might meke people more hesitant to merge PRs when they **should** be also backported, because rather than clicking a button and setting a label, it requires them to actually backport the change and possibly resolve conflicts (which so far was release manager's job).

But also yes, it allows to more easily gate and make sure that what's get "backported" is verified and approved by someone else, it's not an individual decision.

I am quite ok with the burden, as long as all committers are aware of that. I thknk if we do not communicate it explicitly and explain that this is "expectation" - we might end up with people having different expectation and not making such PRs - because they will think it will somehow "magically" happen.

Reply:
(and yes cherry-picker was planned for it but I am not sure if this will work with branch protection and how it will look like - and whether we have someone working on it).

Reply:
As far as I remember we didn't have test branches protected before

Reply:
> As far as I remember we didn't have test branches protected before

Correct. This is a change I've leadner from the discussion I linked before - I think we have not discussed it (or at least I can't remember discussion about that in any place).

Reply:
@ephraimbuddy added it here https://github.com/apache/airflow/commit/06d6ecadffe7734e8758cf0067c73fd81d73c355

Proposal was for the PR author to raise it if they want to guarantee inclusion of a bugfix in 2.x -- otherwise it is on the committer who merges it. (Although yeah branch protection with cherry-picking needs approval so pros/cons)

Reply:
Side comment: 

Maybe it's just me but I think `cherry-picking` use is misleading if we have "branch protection" (at least that was what made me cherry-pick directly to v2-10-test. Possibly we should use "back-porting PR" term from now on - because cherry-picking commits directly to branch is "easier" and can be done  in bulk (even 100s of commits) with very little overhead (if no conflicts found). Creating a PR for each commit that you want to back-port and getting them approved, tests passing and merging. is quite a bit more overhead (also it involves breaking the flow  and context switching as it has to wait for CI and approval) - so I'd say "back-porting" is a better way to describe it, even if it starts with creating new branch and running `git cherry-pick -x`.

Also realisticly speaking - I think it will be quite problematic to expect contributor to raise yet another PR to a different branch. Just explaining and guiding contributors to do so and dealing with backporting issues it might cause, might take more time than just creating such backport.

Reply:
>Maybe it's just me but I think cherry-picking use is misleading if we have "branch protection" (at least that was what made me cherry-pick directly to v2-10-test).

Yeah, agreed with cherry-picking and branch-protection issue.

>Also realisticly speaking - I think it will be quite problematic to expect contributor to raise yet another PR to a different branch. Just explaining and guiding contributors to do so and dealing with backporting issues it might cause, might take more time than just creating such backport PR.

Our language on that I think is good enough: "If there are merge conflicts, the committer should add a comment on the original PR, informing the author and asking them
to raise a separate PR against ``v2-10-test`` branch. If this doesn't happen, there is no guarantee that the PR will be part of 2.10.x"

Reply:
Yeah it is . As long as we replace "cherry-picking" in all the places by "backporting the PR against v2-10-test branch" or something. It's all clear (though it needs to be iterated and reminded a few times at devlist and #contributing channel I think - a lot of things is happening and it's easy to miss that important expectation.

Reply:
The problem with "cherry-picking" directly to the test branch without raising a PR is that we would run into the issue of someone force-pushing and removing someone else's commits. If there's another way to make it so we don't raise a PR and don't force push to v2-10-test, that would be better.

Reply:
> The problem with "cherry-picking" directly to the test branch without raising a PR is that we would run into the issue of someone force-pushing and removing someone else's commits.


that is a good point

Reply:
If the only problem is overriding each-other's pushes - Github allows to protect branches and not allow force-push - it would require to add this feature to .asf.yml - similarly as i did with unresolved conversations - because I think i't's not supported now by the infrastructure:

https://github.com/apache/infrastructure-p6/pull/1740

It's rather easy - so if we think it's worth it maybe someone can raise a PR there ? Infra team is happy to accept such changes.

But I think - even if it makes things slower -  it's worth gating it, also if we have a need to bulk-cherry-pick PRs we could potentialy open a "bulk" PR and cherry-pick mutlitple commits (and then move the tag after approval rather than merge it from UI - whicih will keep the history and avoid squashing those commits).

Reply:
Just want to mention that from how the [cherrypicker](https://pypi.org/project/cherry_picker/) works, the branches don't allow direct pushes but PRs of cherry-picked commits: https://github.com/python/cpython/pull/122312.

Reply:
Yeah. Having cherry-picker configured would save a bit of the overhead - looking how it works making the process more applicable even to contributors.

Reply:
Is this something we want to open for discussion in the next call or do we have alignment here?

Reply:
Absolutely

Reply:
updated